agent/src/agentclient.rs

```diff
@@ -3,102 +3,123 @@ use k8s_openapi::api::core::v1::Node;
 use kube::Api;
 use snafu::OptionExt;
 use snafu::ResultExt;
-use std::env;
 use std::process::Command;
 use tokio::time::{sleep, Duration};
 
+use kube::runtime::reflector::Store;
+
 use crate::apiclient::{boot_update, get_os_info, list_available, prepare, update};
 use crate::error::{self, Result};
 use apiserver::api::{CreateBottlerocketNodeRequest, UpdateBottlerocketNodeRequest};
 use models::constants::{APISERVER_SERVICE_NAME, APISERVER_SERVICE_PORT, NAMESPACE};
 use models::node::{
-    node_resource_name, BottlerocketNode, BottlerocketNodeSelector, BottlerocketNodeState,
-    BottlerocketNodeStatus,
+    BottlerocketNode, BottlerocketNodeSelector, BottlerocketNodeState, BottlerocketNodeStatus,
 };
 
 const AGENT_SLEEP_DURATION: Duration = Duration::from_secs(5);
 const NODE_RESOURCE_ENDPOINT: &'static str = "/bottlerocket-node-resource";
+const MAX_ATTEMPTS: i8 = 5;
+const STREAM_SLEEP_DURATION: Duration = Duration::from_secs(10);
 
 #[derive(Clone)]
 pub struct BrupopAgent {
     k8s_client: kube::client::Client,
-    node_selector: Option<BottlerocketNodeSelector>,
+    brn_reader: Store<BottlerocketNode>,
+    node_reader: Store<Node>,
+    associated_node_name: String,
+    associated_bottlerocketnode_name: String,
 }
 
 impl BrupopAgent {
-    pub fn new(k8s_client: kube::client::Client) -> Self {
+    pub fn new(
+        k8s_client: kube::client::Client,
+        brn_reader: Store<BottlerocketNode>,
+        node_reader: Store<Node>,
+        associated_node_name: String,
+        associated_bottlerocketnode_name: String,
+    ) -> Self {
         BrupopAgent {
             k8s_client,
-            node_selector: None,
+            brn_reader,
+            node_reader,
+            associated_node_name,
+            associated_bottlerocketnode_name,
         }
     }
 
-    pub async fn check_node_custom_resource_exists(&mut self) -> Result<bool> {
-        let associated_bottlerocketnode_name = node_resource_name(&self.get_node_selector().await?);
+    pub async fn check_node_custom_resource_exists(&self) -> Result<bool> {
         let bottlerocket_nodes: Api<BottlerocketNode> =
             Api::namespaced(self.k8s_client.clone(), NAMESPACE);
 
         Ok(bottlerocket_nodes
```

Comment:
Should I use reflector and query BRN object if node custom resource exists here? The reason why I did not update it because querying `Store` sometime will delay, so sometime when I query `Store`, it return empty to me, which would be recognized  BRN had not been created. To solve this issue, I added retry and sleep to give it more time to make objects data ready. Do you think It's appropriate to use retry and sleep here?

Reply:
If this is called during the event loop, it might not be a great idea to always make the API call.

I would either:
* Check the store. If it's not present in the store, make the API call.
* OR do retries with backoff on the store.

I think option 1 makes the most sense, since it returns more quickly.